### ISSUE TITLE
mstpr-brainbot - Liquidity incentive for depositing liquidity is wrong

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-12-dodo-gsp-judging#110]
mstpr-brainbot

high

# Liquidity incentive for depositing liquidity is wrong

## Summary
When users deposits tokens to mint shares the incentive for providing the scarce token is not existing. Contrary, the incentive is given to the most amounted token which incentives users to deposit the token that has more in the pool. 
## Vulnerability Detail
Let's consider an example where a pool initially has 100-100 tokens at time t=0, with a total of 100 shares in totalSupply, representing perfect equilibrium.

After several trades, the token balances in the pool change to 20-180, indicating that the base token is scarce while the quote token is in abundance within the pool. This situation should ideally incentivize users to deposit base tokens in order to receive more LP tokens since the pool is not in equilibrium. However, this doesn't occur due to the way the share calculation operates when a user deposits tokens into the pool:
```solidity
uint256 baseInputRatio = DecimalMath.divFloor(baseInput, baseReserve);
uint256 quoteInputRatio = DecimalMath.divFloor(quoteInput, quoteReserve);
uint256 mintRatio = quoteInputRatio < baseInputRatio ? quoteInputRatio : baseInputRatio;
// The shares will be minted to user
shares = DecimalMath.mulFloor(totalSupply, mintRatio);
```
As observed, the lowest ratio is chosen. Considering we had 20-180 tokens in the pool, if a user intends to deposit a total of 100 tokens, what would be the optimal deposit strategy? Depositing 90 base and 10 quote tokens would increase the base tokens and move the pool towards equilibrium. Let's see how many shares the user would receive by depositing 90-10:

mintRatio = min(90/20, 10/180)
= 10/180

10/180 * 100 = 5.55 shares.

Now, what if the user deposits liquidity in reverse, with 10 base tokens and 90 quote tokens?

mintRatio = min(10/20, 90/180)
= 1/2

1/2 * 100 = 50 shares!

Thus, depositing 10-90, despite the pool having significantly more quote tokens, would grant the user nearly 10 times more shares. Users incentivized to provide liquidity deviate the pool further from its balanced state, which contradicts the pool's equilibrium!

Here a coded PoC:
```solidity
function test_DepositInFavourOfThePool() external {
        // tapir deposits 100-100 to initiate the pool
        // assume the pool has 20-180, dont want to do sells to further complexify the issue
        // so I'll start the pool with 20-180 from start. 
        vm.startPrank(tapir);
        dai.safeTransfer(address(gsp), 20 * 1e18);
        usdc.transfer(address(gsp), 180 * 1e6);
        gsp.buyShares(tapir);
        
        // deposit 90 base and 10 quote, which is better for the pools equilibrum
        dai.safeTransfer(address(gsp), 90 * 1e18);
        usdc.transfer(address(gsp), 10 * 1e6);
        (uint256 shares, , ) = gsp.buyShares(tapir);

        console.log("Tapir received shares", shares);
        console.log("Tapir received shares human readable form", shares / 1e18);
    }

    function test_DepositNotInFavourOfThePool() external {
        // tapir deposits 100-100 to initiate the pool
        // assume the pool has 20-180, dont want to do sells to further complexify the issue
        // so I'll start the pool with 20-180 from start. 
        vm.startPrank(tapir);
        dai.safeTransfer(address(gsp), 20 * 1e18);
        usdc.transfer(address(gsp), 180 * 1e6);
        gsp.buyShares(tapir);
        
        // deposit 10 base and 90 quote, which is worse for the pools equilibrum
        dai.safeTransfer(address(gsp), 10 * 1e18);
        usdc.transfer(address(gsp), 90 * 1e6);
        (uint256 shares, , ) = gsp.buyShares(tapir);

        console.log("Tapir received shares", shares);
        console.log("Tapir received shares human readable form", shares / 1e18);
    }
```

Test result and logs:
<img width="501" alt="image" src="https://github.com/sherlock-audit/2023-12-dodo-gsp-mstpr/assets/120012681/334dba64-2a80-4a41-9344-80670aa8f438">

## Impact
High, since the pool is never incentivized to be in equilibrium when users provide liquidity. Depositing the scarce token should give the user a bonus on shares since the user helps the pool to reach its equilibrium.
## Code Snippet
https://github.com/sherlock-audit/2023-12-dodo-gsp/blob/af43d39f6a89e5084843e196fc0185abffe6304d/dodo-gassaving-pool/contracts/GasSavingPool/impl/GSPFunding.sol#L31-L82
## Tool used

Manual Review

## Recommendation
Change the mintRatio calculation

**Comment by Banditx0x** [comment]
Isn't a liquidity deposit always in the same ratio as the underlying pool in every AMM such as uniswap, curve etc? Liquidity deposits aren't meant to have "rebalance" the pool.

**Comment by Banditx0x** [comment]
Escalate

see comment above

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> see comment above

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by nevillehuang** [comment]
@Banditx0x Yup understood this point, my concern is if it will allow a potential user to gain more than intended LP shares due to the logic of mintRatio i.e the minimum is always taken, and this issue possibly have shown that

**Comment by chinmay-farkya** [comment]
I agree with @Banditx0x. Additional comments :
1. Usually, the equilibrium the watson is talking about is the 20-180 itself (because it is assumed to be driven by market) and not the ideal 100 -100
2. The user needs to be incentivized to maintain the current balance of the pool (as per uni v2 design)

But there is a special condition here, gsp is only talking about stablecoin to stablecoin pools, so it might make sense to incentivize user to bring balance of pool to the 100-100 ideal scenario (such that it shall have equivalent amounts). The inherent assumption here is that the price of all stablecoins will always be the same (which is not true in case any of those depegs). 

imo this issue is still invalid because pricing shall be left to the market in all cases (thus the pool will be good even if one or both assets depeg).

**Comment by nevillehuang** [comment]
@chinmay-farkya, I agree with your points, and is also what was brought up based on my discussions with @cergyk and the watson that submitted this issue.

My concern is on whether due to the logic of shares accredited to user based on mintRatio, that there is a way that the shares obtained on imbalanced liquidity provision would allow the liquidity provider to profit more than he is incentivized to maintain the balance of the pool.

**Comment by chinmay-farkya** [comment]
> that there is a way that the shares obtained on imbalanced liquidity provision would allow the liquidity provider to profit

I don't think this can happen. User gets the minimum in case of imbalanced LP so when he redeems, he gets back the balanced quantities based on the state of the pool at that time. I guess this is similar to Uniswap v2. You can read the reasoning here : [univ2book](https://www.rareskills.io/post/uniswap-v2-mint-and-burn#:~:text=The%20fact%20that%20the%20user%20will%20get%20the%20worse%20of%20the%20two%20ratios)

So based on the reasoning in above text, I don't see the user profiting.

**Comment by mstpr** [comment]
@chinmay-farkya @Banditx0x 
Comparing UniV2 and a stable pool is technically wrong. UniV2 is not a stable pool in its nature so the liquidity to be added has to be in same ratios, pool will not let you add arbitrary liquidity amounts in the first place. However, this is a "stable" pool and new liquidity adds are never incentivized and it is incentivized in opposite ways which is not the correct behaviour of a stable pool. Curve/Balancer and every other stable pool has a "bonus" when the liquidity to be added makes the pool approach to its equilibrium. 

Pools does not only arbed its way to equilibrium by swaps but with add liquidity too. 

When the pool is 90%-10% in a "stable" pool it clearly means that one asset is depegged. In such pool state, user bringing the non scarce asset will mint more LP tokens and will be able to claim the more of the LP tokens in the scarce side which this behaviour will make the pool even worse in equilibrium

**Comment by mstpr** [comment]
Also, pool is not even following "deposit with the same ratios" pool follows "deposit more of the token that is in the pool for more LP"

**Comment by mstpr** [comment]
> Isn't a liquidity deposit always in the same ratio as the underlying pool in every AMM such as uniswap, curve etc? Liquidity deposits aren't meant to have "rebalance" the pool.

no, it doesn't. UniV2 you can only add with same ratios because it is not meant to be a stable pool. Both curve/balancer and any stable pool if you deposit more of the scarce token you get more LP. What's more important is if you deposit the other token that pool has plenty you get lesser LP tokens which will make you not deposit more of that token to further imbalance the pool. In DODO GSP it is the very opposite, if you bring more of the plenty asset you get more LP. With those LP's you can claim more portion of the scarce token which is ideally "depegged" in a stable pool.

**Comment by mstpr** [comment]
Also, deposits to a pool indeed has to be in same ratios. However, if you allow the users to deposit arbitrary amounts like curve/balancer/dodogsp you must need to give them correct LP amounts. 

Pool is perfect balanced, 50-50 there are 50A 50B tokens deposited by Alice only.

pool is 52-48, there are 52A 48B tokens for Alice.

I deposit more of the A token (cheaper) to get more LP (more than Alice, and cheaper then Alice)

Once I do that pool is even more imbalanced and an arber immediately arbs it back to 52-48% ratios (current market ratios)

I immediately withdraw my LP pocket more than Alice.

**Comment by 10xhash** [comment]
I think the issue should be invalid because the only issue here would be that the team has decided to implement an algorithm in which the price rebalance occurs purely via swaps and not via liquidity deposits. The example shows the loss a user is making if depositing assets out of ratio.

**Comment by 10xhash** [comment]
> @Banditx0x Yup understood this point, my concern is if it will allow a potential user to gain more than intended LP shares due to the logic of mintRatio i.e the minimum is always taken, and this issue possibly have shown that

But there are no gained LP tokens here, in example 1 it is pure loss, while example 2 is a deposit in correct ratio.

**Comment by 10xhash** [comment]
> Also, deposits to a pool indeed has to be in same ratios. However, if you allow the users to deposit arbitrary amounts like curve/balancer/dodogsp you must need to give them correct LP amounts.
> 
> Pool is perfect balanced, 50-50 there are 50A 50B tokens deposited by Alice only.
> 
> pool is 52-48, there are 52A 48B tokens for Alice.
> 
> I deposit more of the A token (cheaper) to get more LP (more than Alice, and cheaper then Alice)
> 
> Once I do that pool is even more imbalanced and an arber immediately arbs it back to 52-48% ratios (current market ratios)
> 
> I immediately withdraw my LP pocket more than Alice.

How would this be valid? If the asset ratio is 52:48, unless you deposit more than 48B tokens, you are not going to obtain more LP tokens than Alice. The amount of LP tokens that you get minted is limited by the amount of B tokens that you deposit. If you are depositing in a ratio >52:48, you would actually be making a loss on the excess of token A that has been deposited. If the arber arbs back to 52:48, wouldn't it actually make Alice obtain more A and B tokens than she originally had before you made the deposit.

**Comment by chinmay-farkya** [comment]
> If you are depositing in a ratio >52:48, you would actually be making a loss on the excess of token A

Yes, and afaiu the watson is suggesting that this shouldn't be the case for stable pools, and that users shall get LPs as the maximums of the two mint ratios

**Comment by Banditx0x** [comment]
@mstpr the distinction between a stable pool and non-stable pool doesn't indicate a need for liquidity deposits to rebalance the pool. Take, for example, the stable pools for Uniswap v3. These are the same as "normal"/non-stable pools, except the LP fee and tickSpacing is typically lower. Additionally, other stable AMM's like curve don't have the same balance of both tokens in their stable pools. 

> @Banditx0x Yup understood this point, my concern is if it will allow a potential user to gain more than intended LP shares due to the logic of mintRatio i.e the minimum is always taken, and this issue possibly have shown that

The issue doesn't show that though. The LP'er gets less when the pool is deviated from the correct price, which is true of almost all AMM's. That's why slippage parameters are required for depositing liquidity.

**Comment by nevillehuang** [comment]
Thanks for all the clarification watsons, based on the discussions, I believe this is a non-issue, given the incentivization is adequate to promote rebalancing of pool.

**Comment by mstpr** [comment]
@Banditx0x, I completely agree with you. I think the primary goal for any LP is to add liquidity in the current ratios. If you balance the pool with your LP, you might get some bonuses, but this is more of a perk than a requirement, so what I mentioned earlier is not a necessity but rather a design choice.

However, I've had another thought on this topic.

In Curve/Balancer, you can deposit any custom amounts, and the ideal liquidity to be added is the actual ratio of the pool (if the pool is 60-40, then add liquidity in a 60-40 ratio). If you add liquidity to help the pool come closer to 50-50, you get a bonus, and if not, you get a penalty.

In UniswapV2/V3 type pools, you can only deposit a certain amount of tokens, not custom amounts. Depending on the pool's balances, if you deposit 10 tokenA, then you must deposit, say, 15 tokenB; you can't deposit 14 or 16 tokenB, you need exactly 15. This is because adding liquidity should never change the price of the pool. In Curve and Balancer, depositing such custom liquidity would indeed change the price if you change the pool's ratios; hence, you get the penalty or bonus. In UniswapV2/V3, we don't have that worry because you won't be able to do that by adding liquidity.

So, back to DODO. Obviously, if the pool has 60-40 balances and 40 total supply, then the ratio of what I should deposit needs to be 3-2 (3 tokenA and 2 tokenB). If I deposit 60 tokenA and 40 tokenB, I get 40 LP, and the pool now has 120 tokenA and 80 tokenB, so I didn't change the price (spot price).

If I deposit 80 tokenA and 40 tokenB, I get 40 LP again, and the pool now has 140 tokenA and 80 tokenB, and I did change the price. Did I have any incentive to do this? No(?), I basically paid 20 more tokenA for the same LP amount. What did I achieve? I made tokenB more expensive in the pool, but can I take advantage of this? Not sure; I'll do some work on this.

TLDR: In Uniswap type pools, you can't add customized amounts of liquidity; your LP amount is fixed with your amounts. In Curve-type stable pools, you can add customized liquidity, but you can incur penalties or bonuses in LP amounts. DODO allows you to deposit both tokens in any custom ratio, like a stable pool, and does not have bonuses or penalties in LPs, which makes the DODO pools' price changeable by adding liquidity.

**Comment by mstpr** [comment]
> @Banditx0x, I completely agree with you. I think the primary goal for any LP is to add liquidity in the current ratios. If you balance the pool with your LP, you might get some bonuses, but this is more of a perk than a requirement, so what I mentioned earlier is not a necessity but rather a design choice.
> 
> However, I've had another thought on this topic.
> 
> In Curve/Balancer, you can deposit any custom amounts, and the ideal liquidity to be added is the actual ratio of the pool (if the pool is 60-40, then add liquidity in a 60-40 ratio). If you add liquidity to help the pool come closer to 50-50, you get a bonus, and if not, you get a penalty.
> 
> In UniswapV2/V3 type pools, you can only deposit a certain amount of tokens, not custom amounts. Depending on the pool's balances, if you deposit 10 tokenA, then you must deposit, say, 15 tokenB; you can't deposit 14 or 16 tokenB, you need exactly 15. This is because adding liquidity should never change the price of the pool. In Curve and Balancer, depositing such custom liquidity would indeed change the price if you change the pool's ratios; hence, you get the penalty or bonus. In UniswapV2/V3, we don't have that worry because you won't be able to do that by adding liquidity.
> 
> So, back to DODO. Obviously, if the pool has 60-40 balances and 40 total supply, then the ratio of what I should deposit needs to be 3-2 (3 tokenA and 2 tokenB). If I deposit 60 tokenA and 40 tokenB, I get 40 LP, and the pool now has 120 tokenA and 80 tokenB, so I didn't change the price (spot price).
> 
> If I deposit 80 tokenA and 40 tokenB, I get 40 LP again, and the pool now has 140 tokenA and 80 tokenB, and I did change the price. Did I have any incentive to do this? No(?), I basically paid 20 more tokenA for the same LP amount. What did I achieve? I made tokenB more expensive in the pool, but can I take advantage of this? Not sure; I'll do some work on this.
> 
> TLDR: In Uniswap type pools, you can't add customized amounts of liquidity; your LP amount is fixed with your amounts. In Curve-type stable pools, you can add customized liquidity, but you can incur penalties or bonuses in LP amounts. DODO allows you to deposit both tokens in any custom ratio, like a stable pool, and does not have bonuses or penalties in LPs, which makes the DODO pools' price changeable by adding liquidity.

oh ok....

so in curve,
if the pool is 60-40
I deposit 60-40, say, I get 1 LP token. 
I deposit 65-40, say, I get 1.02 LP token! In DODO I get exactly 1 LP token so actually DODO's penalty is more than curve. 

if I deposit 1-20 making the balances 61-60 helping the pool balance I get 1/60 ratio LP tokens which again I have the penalty but that's ok as you said, bonus is not mandatory. 

So, DODO has no bonus but strict penalties then curve which is ok. Though, if you think I am wrong please correct me, this is an interesting topic for me!

@Czar102 @nevillehuang I agree with escalations, this is invalid and a design choice.

**Comment by Czar102** [comment]
Agree with the escalation, it's a valid and common design. There is no loss of funds here. You can execute a swap and provide liquidity and performing such action is incentivized by the invariant.

The current behavior also means that it's possible to frontrun a liquidity addition to make the number of shares lower, hence increasing the value of existing shares. Has this separate issue been reported?

**Comment by Czar102** [comment]
Result:
Invalid
Unique

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [banditx0x](https://github.com/sherlock-audit/2023-12-dodo-gsp-judging/issues/110/#issuecomment-1890185111): accepted

**Comment by chinmay-farkya** [comment]
> Agree with the escalation, it's a valid and common design. There is no loss of funds here. You can execute a swap and provide liquidity and performing such action is incentivized by the invariant.
> 
> The current behavior also means that it's possible to frontrun a liquidity addition to make the number of shares lower, hence increasing the value of existing shares. Has this separate issue been reported?

Hey @Czar102 , the "separate issue" you are referring to : an issue along those lines has been reported. Here it is : https://github.com/sherlock-audit/2023-12-dodo-gsp-judging/issues/68

Idk if this has to be accepted, but @nevillehuang did point out in discord that router design was not known to the auditors. I'll let you decide.

**Comment by Czar102** [comment]
The judgments are correct. Good to see the issue being submitted. I wasn't sure if this function is to be interacted with through a router.
Thank you for the clarification @chinmay-farkya.

**Comment by nevillehuang** [comment]
Oops slipped my mind @Czar102 my bad on the wrong info relayed.
